### PR TITLE
SCP-2087: Update context definition

### PR DESCRIPTION
**Pull request by j-mueller**
* Change the `TxInfo` type from our provisional version to [the real version ](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/alonzo/impl/src/Cardano/Ledger/Alonzo/FakePlutus.hs) that is going to be produced by the Goguen node
* Notable changes
  * Address has staking credentials
  * `TxInfo` has staking withdrawal data
  * Fee inputs are separated from other inputs in `TxInfo`
  * `ValidatorCtx` and `ScriptCtx` have been replaced by a single `ScriptContext` type with a `scriptPurpose` field

Note that the emulator doesn't support staking so the staking related fields will always be empty or `Nothing`.

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs [hunk 1]
```diff
@@ -92,20 +93,20 @@ machineAddress = scriptAddress . validatorInstance
 -- | Turn a state machine into a validator script.
 mkValidator :: forall s i. (PlutusTx.IsData s) => StateMachine s i -> ValidatorType (StateMachine s i)
 mkValidator (StateMachine step isFinal check) currentState input ptx =
-    let vl = txInInfoValue (findOwnInput ptx)
+    let vl = maybe (error ()) (txOutValue . txInInfoResolved) (findOwnInput ptx)
```
Comment (michaelpj) [inline plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:96]: Hmm, this is vaguely unsatisfactory, but not much we can do about it. We're exposing the error case from `findOwnInput` to users, which I guess is okay.

File: plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs [hunk 2]
```diff
@@ -11,50 +11,77 @@
...
 import           Plutus.V1.Ledger.Crypto
-import           Plutus.V1.Ledger.Orphans  ()
+import           Plutus.V1.Ledger.Orphans    ()
 import           Plutus.V1.Ledger.Scripts
 
--- | A payment address using a hash as the id.
-data Address = PubKeyAddress PubKeyHash
-    | ScriptAddress ValidatorHash
+-- | Address with two kinds of credentials, normal and staking
+data Address = Address{ addressCredential :: !Credential, addressStakingCredential :: !(Maybe StakingCredential)}
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs:38]: let's not include the bangs, I don't think they do anything and it'll make people think that they do

File: plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs [hunk 3]
```diff
@@ -11,50 +11,77 @@
...
     deriving stock (Eq, Ord, Show, Generic)
     deriving anyclass (ToJSON, FromJSON, ToJSONKey, FromJSONKey, Serialise, Hashable, NFData)
 
 instance Pretty Address where
-    pretty (PubKeyAddress pkh) = "PubKeyAddress:" <+> pretty pkh
-    pretty (ScriptAddress vh)  = "ScriptAddress:" <+> pretty vh
+    pretty (Address cred stakingCred) = pretty cred <+> pretty stakingCred
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs:43]: Should we do something prettier than this?

File: plutus-ledger-api/src/Plutus/V1/Ledger/Credential.hs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+instance PlutusTx.Eq StakingCredential where
+    StakingHash l == StakingHash r = l PlutusTx.== r
+    StakingPtr a b c == StakingPtr a' b' c' =
+        a PlutusTx.== a'
+        PlutusTx.&& b PlutusTx.== b'
+        PlutusTx.&& c PlutusTx.== c'
+    _ == _ = False
+
+data Credential
+  = PubKeyCredential !PubKeyHash
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Credential.hs:55]: remove the bangs

File: plutus-ledger-api/src/Plutus/V1/Ledger/Credential.hs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+instance PlutusTx.Eq StakingCredential where
+    StakingHash l == StakingHash r = l PlutusTx.== r
+    StakingPtr a b c == StakingPtr a' b' c' =
+        a PlutusTx.== a'
+        PlutusTx.&& b PlutusTx.== b'
+        PlutusTx.&& c PlutusTx.== c'
+    _ == _ = False
+
+data Credential
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Credential.hs:54]: haddock

File: plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs [hunk 6]
```diff
@@ -321,28 +319,11 @@ pubKeyTxIn r = TxIn r ConsumePublicKeyAddress
...
-        PayToScript ds -> "PayToScript:" <+> pretty ds
-        PayToPubKey    -> "PayToPubKey"
-
--- | A transaction output, consisting of a target address, a value, and an output type.
+-- | A transaction output, consisting of a target address, a value, and optionally a datum hash.
 data TxOut = TxOut {
-    txOutAddress :: !Address,
-    txOutValue   :: !Value,
-    txOutType    :: !TxOutType
+    txOutAddress   :: !Address,
```
Comment (michaelpj) [inline plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs:324]: get rid of the bangs. Although maybe they help us with the off-chain version, in which case disregard all these comments...

**Review by michaelpj — APPROVED** [review]
LGTM!

File: plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs:38]: Good point. I'll remove them.

File: plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-contract/src/Plutus/Contract/StateMachine/OnChain.hs:96]: Yeah I pushed it out to here, previously `findOwnInput` just failed.

File: plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs (hunk 6 above)
Comment (j-mueller) [reply plutus-ledger-api/src/Plutus/V1/Ledger/Tx.hs:324]: You get a warning about a strict field not being initialised if you forget it... not sure if that's worth it. I doubt it has any performance implications for us (although I never really looked into it, so I could be wrong)

**Comment by michaelpj** [comment]
Tutorials need updating (hooray for compiling our examples!)

File: plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs (hunk 3 above)
Comment (j-mueller) [reply plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs:43]: I need to check what it actually looks like :D maybe I should add an "Address:" at the beginning.

Comment (j-mueller) [reply plutus-ledger-api/src/Plutus/V1/Ledger/Address.hs]: ok I made it a little nicer.

**Comment by kk-hainq** [comment]
@j-mueller Is there any documentation either public or private that specifies/tracks these changes? If not I can update these details to the UTXO specifications through #3707.

**Comment by michaelpj** [comment]
The formal version of this is (I think still in progress) in the Alonzo spec in https://github.com/input-output-hk/cardano-ledger-specs. The EUTXO doc in this repository is at this point an out-of-date design doc... we may delete it.

**Comment by kk-hainq** [comment]
Yeah, I'm aware of the ones in `cardano-ledger-specs` too. But I think they are a bit too involved for an average Plutus developer who just wants to focus on Smart Contracts. What we've been looking for is clean formal definitions of all the concepts that those smart contract developers need, preferably in the scope of this repository. We still have to jump around to study things at the moment.

A bit more context but we started this documentation effort because we haven't had the best experience training Plutus engineers. We'd love every one of them to read every single thing too, but the steps are still cloudy, with a decent gap between the current Plutus docs and the Ledger docs.

**Comment by michaelpj** [comment]
The best thing to refer to in those circumstances is probably just the type definitions.

**Comment by michaelpj** [comment]
We know we need to write more documentation :(

**Comment by kk-hainq** [comment]
We know you're all overloaded at the moment so if you need anything just ping us.
For documentation, we want to do EUTXO scripting next. Should we update the technical report, the EUTXO specifications, or just write a new markdown?

**Pull request merged**